src/cargo/ops/cargo_install.rs

```diff
@@ -316,40 +329,77 @@ fn install_one(
         );
     }
 
-    // Helper for --no-track flag to make sure it doesn't overwrite anything.
-    let no_track_duplicates = || -> CargoResult<BTreeMap<String, Option<PackageId>>> {
-        let duplicates: BTreeMap<String, Option<PackageId>> = exe_names(&pkg, &opts.filter)
-            .into_iter()
-            .filter(|name| dst.join(name).exists())
-            .map(|name| (name, None))
-            .collect();
-        if !force && !duplicates.is_empty() {
-            let mut msg: Vec<String> = duplicates
-                .iter()
-                .map(|(name, _)| format!("binary `{}` already exists in destination", name))
-                .collect();
-            msg.push("Add --force to overwrite".to_string());
-            bail!("{}", msg.join("\n"));
-        }
-        Ok(duplicates)
-    };
-
     // WARNING: no_track does not perform locking, so there is no protection
     // of concurrent installs.
     if no_track {
         // Check for conflicts.
-        no_track_duplicates()?;
+        no_track_duplicates(&pkg, opts, &dst, force)?;
     } else if is_installed(&pkg, config, opts, &rustc, &target, root, &dst, force)? {
         let msg = format!(
             "package `{}` is already installed, use --force to override",
             pkg
         );
         config.shell().status("Ignored", &msg)?;
-        return Ok(false);
+        return Ok(None);
     }
 
+    Ok(Some(pkg))
+}
+
+fn no_track_duplicates(
+    pkg: &Package,
+    opts: &ops::CompileOptions,
+    dst: &Path,
+    force: bool,
+) -> CargoResult<BTreeMap<String, Option<PackageId>>> {
+    // Helper for --no-track flag to make sure it doesn't overwrite anything.
+    let duplicates: BTreeMap<String, Option<PackageId>> = exe_names(&pkg, &opts.filter)
+        .into_iter()
+        .filter(|name| dst.join(name).exists())
+        .map(|name| (name, None))
+        .collect();
+    if !force && !duplicates.is_empty() {
+        let mut msg: Vec<String> = duplicates
+            .iter()
+            .map(|(name, _)| format!("binary `{}` already exists in destination", name))
+            .collect();
+        msg.push("Add --force to overwrite".to_string());
+        bail!("{}", msg.join("\n"));
+    }
+    Ok(duplicates)
+}
+
+fn install_one<'cfg>(
+    config: &Config,
+    root: &Filesystem,
+    source_id: SourceId,
+    vers: Option<&str>,
+    opts: &ops::CompileOptions,
+    force: bool,
+    no_track: bool,
+    pkg: Package,
+) -> CargoResult<()> {
     config.shell().status("Installing", &pkg)?;
 
+    let dst = root.join("bin").into_path_unlocked();
+    let (mut ws, rustc, target) = make_ws_rustc_target(config, opts, &source_id, pkg.clone())?;
```

Comment:
Called this `make_ws_rustc_target` twice - once in the validation step (`determine_package` and once here).
We could consider funneling the result through from the first call into args here, but I figured it would be easier to just call it again.

Reply:
I think yeah let's try to funnel the result from the previous step into this one, workspace construction can be somewhat nontrivial since it involves parsing TOML files, so minimizing the calls there I think makes sense. I think this would also handle the "overwrite the `pkg` var" comment from above as well?

If you'd like I think it's also fine to have some other refactorings here to avoid "pass all the params all the time" and have, for example, a struct with state in it that has methods. I'll leave that up to you though

Reply:
Great idea! Looks a lot nicer now with `install_one` as a method on `InstallablePackage`. Definitely results in only calling `make_ws_rustc_target` once.

Might be a bit difficult to code review as a bigish change now. Let me know if you have any factoring ideas that might make it easier on you.